src/mon/MonMap.h

```diff
@@ -19,44 +19,176 @@
+  mon_feature_t get_required_features() const {
+    return (persistent_features | optional_features);
+  }
+
+  void sanitize_mons(map<string,entity_addr_t>& o) {
+
+    // if mon_info is populated, it means we decoded a map encoded
+    // by someone who understands the new format (i.e., is able to
+    // encode 'mon_info'). This means they must also have provided
+    // a properly populated 'mon_addr' (which we have dropped with
+    // this patch), 'o' being the contents of said map. In this
+    // case, 'o' must have the same number of entries as 'mon_info'.
+    //
+    // Also, for each entry in 'o', there has to be a matching
+    // 'mon_info' entry, properly populated with a name and a matching
+    // 'public_addr'.
+    //
+    // OTOH, if 'mon_info' is not populated, it means the one that
+    // originally encoded the map does not know the new format, and
+    // 'o' will be our only source of info about the monitors in the
+    // cluster -- and we will use it to populate our 'mon_info' map.
+
+    bool has_mon_info = false;
+    if (mon_info.size() > 0) {
+      assert(o.size() == mon_info.size());
+      has_mon_info = true;
+    }
+
+    for (map<string, entity_addr_t>::const_iterator p = o.begin();
+         p != o.end();
+         ++p) {
+
+      // make sure the info we have is accurate
+      if (has_mon_info) {
+        assert(mon_info.count(p->first));
+        assert(mon_info[p->first].name == p->first);
+        assert(mon_info[p->first].public_addr == p->second);
+        continue;
+      }
+
+      mon_info_t &m = mon_info[p->first];
+      m.name = p->first;
+      m.public_addr = p->second;    
+    }
+  }
 
   void calc_ranks() {
-    rank_name.resize(mon_addr.size());
-    rank_addr.resize(mon_addr.size());
-    addr_name.clear();
-    for (map<string,entity_addr_t>::iterator p = mon_addr.begin();
-	 p != mon_addr.end();
-	 ++p) {
-      assert(addr_name.count(p->second) == 0);
-      addr_name[p->second] = p->first;
+
+    ranks.resize(mon_info.size());
+    addr_mons.clear();
+
+    // Used to order entries according to public_addr, because that's
+    // how the ranks are expected to be ordered by. We may expand this
+    // later on, according to some other criteria, by specifying a
+    // different comparator.
+    //
+    // Please note that we use a 'set' here instead of resorting to
+    // std::sort() because we need more info than that's available in
+    // the vector. The vector will thus be ordered by, e.g., public_addr
+    // while only containing the names of each individual monitor.
+    // The only way of achieving this with std::sort() would be to first
+    // insert every mon_info_t entry into a vector 'foo', std::sort() 'foo'
+    // with custom comparison functions, and then copy each invidual entry
+    // to a new vector. Unless there's a simpler way, we don't think the
+    // added complexity makes up for the additional memory usage of a 'set'.
+    set<mon_info_t, rank_cmp> tmp;
```

Comment:
- can we move the implementation of this method into MonMap.cc to speed up the compilation?
- could also define the comparator using `WRITE_CMP_OPERATORS_2`.